### PR TITLE
Passthrough Mini DB

### PR DESCRIPTION
**Pull request by valaises**
Changelist:
* created mini_db for passthrough models
* resolving passthrough models to up-to-date versions
* new field in caps: n_ctx_rewrite (specifying context for each running model)

### CODE REVIEW

File: known_models_db/refact_known_models/passthrough.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        "filter_caps": ["chat"],
+    },
+    # gpt-3.5-turbo-0125 is already available, but no support from litellm yet.
+    "gpt-3.5-turbo": {
+        "provider": "openai",
+        "model_path": "Xenova/gpt-3.5-turbo-16k",
+        "resolve_as": "gpt-3.5-turbo-1106",
+        "T": 16_000,
+        "T_out": 4096,
+        "filter_caps": ["chat"],
```
Comment (valaises) [inline known_models_db/refact_known_models/passthrough.py:20]: leave empty

File: known_models_db/refact_known_models/passthrough.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        "provider": "openai",
+        "model_path": "Xenova/gpt-4",
+        "resolve_as": "gpt-4-0125-preview",
+        "T": 128_000,
+        "T_out": 4096,
+        "filter_caps": ["chat"],
+    },
+    # gpt-3.5-turbo-0125 is already available, but no support from litellm yet.
+    "gpt-3.5-turbo": {
+        "provider": "openai",
```
Comment (valaises) [inline known_models_db/refact_known_models/passthrough.py:15]: add backend

File: refact_utils/finetune/utils.py [hunk 3]
```diff
@@ -97,12 +98,10 @@ def add_result(k: str, model_dict: Dict):
             result['chat'].append(k)
 
     if data.get("openai_api_enable"):
-        add_result("gpt-3.5-turbo", {'has_chat': True})
-        add_result("gpt-4", {'has_chat': True})
+        [add_result(k, {'has_chat': True}) for k, v in passthrough_mini_db.items() if v.get('provider') == 'openai' and 'chat' in v.get('filter_caps', [])]
```
Comment (valaises) [inline refact_utils/finetune/utils.py:101]: make a function instead

File: refact_webgui/webgui/selfhost_model_resolve.py [hunk 4]
```diff
@@ -50,3 +51,14 @@ def _family(model: str) -> str:
             return have_model, ""
     else:
         return "", f"model \"{model_name}\" is not loaded (3)"
+
+
+def resolve_model_context_size(model_name: str) -> Optional[int]:
```
Comment (valaises) [inline refact_webgui/webgui/selfhost_model_resolve.py:56]: use model_assigner

Comment (valaises) [reply refact_webgui/webgui/selfhost_model_resolve.py:56]: no imports from known_models_db!

File: refact_webgui/webgui/selfhost_queue.py [hunk 5]
```diff
@@ -53,11 +54,11 @@ def models_available(self, force_read: bool = False) -> List[str]:
             for model in j["model_assign"]:
                 self._models_available.append(model)
             self._models_available_ts = time.time()
-            if j.get("openai_api_enable", False):
```
Comment (valaises) [inline refact_webgui/webgui/selfhost_queue.py:56]: pass model_assigner to InferenceQueue

File: known_models_db/refact_known_models/passthrough.py (hunk 1 above)
Comment (valaises) [reply known_models_db/refact_known_models/passthrough.py:20]: mb remove?

File: known_models_db/refact_known_models/passthrough.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        "T": 200_000,
+        "T_out": 4096,
+        "filter_caps": ["chat"],
+    },
+    "claude-2.1": {
+        "provider": "anthropic",
+        "model_path": "Xenova/claude-tokenizer",
+        "resolve_as": "claude-2.1",
+        "T": 100_000,
+        "T_out": 4096,
```
Comment (valaises) [inline known_models_db/refact_known_models/passthrough.py:51]: pass it to litellm as param

File: known_models_db/refact_known_models/passthrough.py (hunk 1 above)
Comment (valaises) [reply known_models_db/refact_known_models/passthrough.py:20]: Cannot remove. It's used in running_models_and_loars at finetune/utils.py

File: refact_utils/finetune/utils.py [hunk 7]
```diff
@@ -96,13 +96,18 @@ def add_result(k: str, model_dict: Dict):
         if model_dict.get('has_chat'):
             result['chat'].append(k)
 
+    def add_results_for_passthrough_provider(provider: str) -> None:
+        [
+            add_result(k, {'has_chat': 'chat' in v.get('filter_caps', [])})
+            for k, v in model_assigner.passthrough_mini_db.items()
+            if v.get('provider') == provider
+        ]
```
Comment (mitya52) [inline refact_utils/finetune/utils.py:104]: * add_results_for_passthrough_provider -> _add_results_for_passthrough_provider
* write without list comprehension
* all passthrough models are about chat so no need in add_result usage here

Comment (mitya52) [reply refact_utils/finetune/utils.py:104]: ...and "chat" caps checking

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 8]
```diff
@@ -14,13 +14,14 @@
 
 from refact_utils.scripts import env
 from refact_utils.finetune.utils import running_models_and_loras
-from refact_webgui.webgui.selfhost_model_resolve import completion_resolve_model
+from refact_webgui.webgui.selfhost_model_resolve import completion_resolve_model, resolve_model_context_size, resolve_tokenizer_name_for_model
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:17]: 3 separate imports please

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 9]
```diff
@@ -299,18 +300,15 @@ def _caps_base_data(self) -> Dict[str, Any]:
             "running_models": [r for r in [*running['completion'], *running['chat']]],
             "code_completion_default_model": code_completion_default_model,
             "code_chat_default_model": code_chat_default_model,
+            "n_ctx_rewrite": {model: t for model in models_available if (t := resolve_model_context_size(model, self._model_assigner))},
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:303]: t -> n_ctx

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 10]
```diff
@@ -570,14 +568,19 @@ async def _chat_completions(self, post: ChatContext, authorization: str = Header
         account = await self._account_from_bearer(authorization)
 
         prefix, postfix = "data: ", "\n\n"
+        model_dict = self._model_assigner.models_db_with_passthrough.get(post.model, {})
+
+        if model_dict.get('backend') == 'litellm' and (model_name := model_dict.get('resolve_as', post.model)) in litellm.model_list:
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:573]: no model_name if backend is not litellm
do not use := operator everywhere

File: refact_webgui/webgui/selfhost_model_resolve.py [hunk 11]
```diff
@@ -50,3 +51,22 @@ def _family(model: str) -> str:
...
+        if max_tokens := model_assigner.passthrough_mini_db[model_name].get('T'):
+            return min(PASSTHROUGH_MAX_TOKENS_LIMIT, max_tokens)
+
+
+def resolve_tokenizer_name_for_model(model_name: str, model_assigner: ModelAssigner) -> Optional[str]:
+    if model_name in model_assigner.models_db:
+        return model_assigner.models_db[model_name].get('model_path')
+
+    if model_name in model_assigner.passthrough_mini_db:
+        return model_assigner.passthrough_mini_db[model_name].get('tokenizer')
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_model_resolve.py:72]: tokenizer...do you mean tokenizer_path?

**Review by mitya52 — CHANGES_REQUESTED** [review]

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 12]
```diff
@@ -15,12 +15,15 @@
...
 from refact_webgui.webgui.selfhost_model_resolve import completion_resolve_model
+from refact_webgui.webgui.selfhost_model_resolve import resolve_model_context_size
+from refact_webgui.webgui.selfhost_model_resolve import resolve_tokenizer_name_for_model
 from refact_webgui.webgui.selfhost_model_resolve import static_resolve_model
 from refact_webgui.webgui.selfhost_queue import Ticket
 from refact_webgui.webgui.selfhost_webutils import log
 from refact_webgui.webgui.selfhost_queue import InferenceQueue
 from refact_webgui.webgui.selfhost_model_assigner import ModelAssigner
 from refact_webgui.webgui.selfhost_login import RefactSession
+from known_models_db.refact_known_models import passthrough_mini_db
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:26]: no need in this import

File: refact_webgui/webgui/selfhost_queue.py [hunk 13]
```diff
@@ -53,11 +55,11 @@ def models_available(self, force_read: bool = False) -> List[str]:
...
-                self._models_available.append('gpt-4')
-            if j.get("anthropic_api_enable", False):
-                self._models_available.append('claude-instant-1.2')
-                self._models_available.append('claude-2.1')
+
+            if j.get("openai_api_enable"):
+                self._models_available.extend(k for k, v in self._model_assigner.passthrough_mini_db.items() if v.get('provider') == 'openai')
+
+            if j.get("anthropic_api_enable"):
+                self._models_available.extend(k for k, v in self._model_assigner.passthrough_mini_db.items() if v.get('provider') == 'anthropic')
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_queue.py:63]: same code, unify it

**Review by mitya52 — CHANGES_REQUESTED** [review]

**Review by mitya52 — APPROVED** [review]

**Pull request merged**